File: server/src/services/expression-review-store.ts
/**
 * 表达方式 AI 审核日志存储（对照 src/learners/expression_review_store.py）。
 *
 * - 日志文件：logs/expression_review/review_logs.json（JSON 数组，整文件重写）
 * - 事件类型：ai_review / manual_rescue（救回通过 review_log_id 关联）
 * - 审核状态（checked/rejected 快照）存于 local_storage 键
 *   `expression_review:{id}`，本模块同样提供读写。
 */

import { randomUUID } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";

import type { FastifyBaseLogger } from "fastify";

import type { LocalStore } from "./local-store.js";

const AI_REVIEW_EVENT = "ai_review";
const MANUAL_RESCUE_EVENT = "manual_rescue";

export type ReviewLogEntry = Record<string, unknown>;

export class ExpressionReviewStore {
  constructor(
    readonly logPath: string,
    private readonly localStore: LocalStore,
    private readonly logger: FastifyBaseLogger,
  ) {}

  static open(rootDir: string, localStore: LocalStore, logger: FastifyBaseLogger): ExpressionReviewStore {
    return new ExpressionReviewStore(
      path.join(rootDir, "logs", "expression_review", "review_logs.json"),
      localStore,
      logger,
    );
  }

  private readEntries(): ReviewLogEntry[] {
    if (!existsSync(this.logPath)) {
      return [];
    }
    try {
      const parsed = JSON.parse(readFileSync(this.logPath, "utf8")) as unknown;
      if (!Array.isArray(parsed)) {
        this.logger.warn("表达方式审核日志格式异常，应为 JSON 数组");
        return [];
      }
      return parsed.filter((entry): entry is ReviewLogEntry => entry !== null && typeof entry === "object");
    } catch (error) {
      this.logger.warn({ error }, "表达方式审核日志 JSON 解析失败，已忽略当前文件");
      return [];
    }
  }

  private appendEntry(entry: ReviewLogEntry): void {
    const entries = this.readEntries();
    entries.push(entry);
    mkdirSync(path.dirname(this.logPath), { recursive: true });
    writeFileSync(this.logPath, `${JSON.stringify(entries, null, 2)}\n`, "utf8");
  }

  private nowIso(): string {
    const offsetMinutes = -new Date().getTimezoneOffset();
    const sign = offsetMinutes >= 0 ? "+" : "-";
    const pad = (value: number, width = 2) => String(value).padStart(width, "0");
    const d = new Date();
    return (
      `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}` +
      `T${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}` +
      `${sign}${pad(Math.abs(offsetMinutes) / 60 | 0)}:${pad(Math.abs(offsetMinutes) % 60)}`
    );
  }

  appendAiReviewLog(input: {
    session_id: string;
    situation: string;
    style: string;
    passed: boolean;
    reason: string;
    source: string;
    expression_id?: number | null;
    error?: string | null;
  }): ReviewLogEntry {
    const entry: ReviewLogEntry = {
      id: randomUUID().replaceAll("-", ""),
      event: AI_REVIEW_EVENT,
      created_at: this.nowIso(),
      expression_id: input.expression_id ?? null,
      session_id: String(input.session_id ?? "").trim(),
      passed: Boolean(input.passed),
      reason: String(input.reason ?? "").trim(),
      situation: String(input.situation ?? "").trim(),
      style: String(input.style ?? "").trim(),
      source: input.source,
    };
    if (input.error) {
      entry.error = String(input.error);
    }
    this.appendEntry(entry);
    return entry;
  }

  appendManualRescueLog(reviewLogId: string, expressionId: number): ReviewLogEntry {
    const entry: ReviewLogEntry = {
      id: randomUUID().replaceAll("-", ""),
      event: MANUAL_RESCUE_EVENT,
      created_at: this.nowIso(),
      review_log_id: String(reviewLogId),
      expression_id: expressionId,
    };
    this.appendEntry(entry);
    return entry;
  }

  private parseCreatedAt(value: unknown): number {
    if (typeof value !== "string" || value === "") {
      return 0;
    }
    const parsed = Date.parse(value);
    return Number.isNaN(parsed) ? 0 : parsed;
  }

  private rescueByReviewId(entries: ReviewLogEntry[]): Map<string, ReviewLogEntry> {
    const rescues = entries
      .filter(
        (entry) =>
          entry.event === MANUAL_RESCUE_EVENT && String(entry.review_log_id ?? "").trim() !== "",
      )
      .sort((a, b) => this.parseCreatedAt(b.created_at) - this.parseCreatedAt(a.created_at));
    const map = new Map<string, ReviewLogEntry>();
    for (const entry of rescues) {
      const reviewLogId = String(entry.review_log_id ?? "").trim();
      if (!map.has(reviewLogId)) {
        map.set(reviewLogId, entry);
      }
    }
    return map;
  }

  private withRescueState(entry: ReviewLogEntry, rescueEntry?: ReviewLogEntry): ReviewLogEntry {
    return {
      ...entry,
      rescued: rescueEntry !== undefined,
      rescued_expression_id: rescueEntry ? (rescueEntry.expression_id ?? null) : null,
      rescued_at: rescueEntry ? (rescueEntry.created_at ?? null) : null,
    };
  }

  getRecentAiReviewLogs(options: {
    limit?: number;
    passed?: boolean | null;
    sessionId?: string | null;
  }): ReviewLogEntry[] {
    const { limit = 50, passed = null, sessionId = null } = options;
    const entries = this.readEntries();
    const rescues = this.rescueByReviewId(entries);
    const normalizedSessionId = String(sessionId ?? "").trim();

    let reviewEntries = entries.filter(
      (entry) => (entry.event ?? AI_REVIEW_EVENT) === AI_REVIEW_EVENT && String(entry.id ?? "").trim() !== "",
    );
    if (passed !== null) {
      reviewEntries = reviewEntries.filter((entry) => Boolean(entry.passed) === passed);
    }
    if (normalizedSessionId) {
      reviewEntries = reviewEntries.filter(
        (entry) => String(entry.session_id ?? "").trim() === normalizedSessionId,
      );
    }
    reviewEntries.sort((a, b) => this.parseCreatedAt(b.created_at) - this.parseCreatedAt(a.created_at));
    return reviewEntries
      .slice(0, Math.max(1, limit))
      .map((entry) => this.withRescueState(entry, rescues.get(String(entry.id))));
  }

  getAiReviewLog(reviewLogId: string): ReviewLogEntry | null {
    const normalizedId = String(reviewLogId ?? "").trim();
    if (!normalizedId) {
      return null;
    }
    const entries = this.readEntries();
    const rescues = this.rescueByReviewId(entries);
    for (const entry of entries) {
      if ((entry.event ?? AI_REVIEW_EVENT) !== AI_REVIEW_EVENT) {
        continue;
      }
      if (String(entry.id ?? "").trim() === normalizedId) {
        return this.withRescueState(entry, rescues.get(normalizedId));
      }
    }
    return null;
  }

  // ------------------------------------------------------------------ 审核状态（local_storage）

  getReviewState(expressionId: number | null): { checked: boolean; rejected: boolean; modified_by: string | null } {
    if (expressionId === null) {
      return { checked: false, rejected: false, modified_by: null };
    }
    const value = this.localStore.get<Record<string, unknown>>(`expression_review:${expressionId}`);
    if (value && typeof value === "object") {
      return {
        checked: Boolean(value.checked),
        rejected: Boolean(value.rejected),
        modified_by: (value.modified_by as string | undefined) ?? null,
      };
    }
    return { checked: false, rejected: false, modified_by: null };
  }

  setReviewState(expressionId: number | null, checked: boolean, rejected: boolean, modifiedBy: string | null): void {
    if (expressionId === null) {
      return;
    }
    this.localStore.set(`expression_review:${expressionId}`, {
      checked,
      rejected,
      modified_by: modifiedBy,
    });
  }
}

File: src/learners/expression_review_store.py
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional

import json
import uuid

from src.common.logger import get_logger
from src.manager.local_store_manager import local_storage

logger = get_logger("expressor")

REVIEW_LOG_PATH = Path("logs/expression_review/review_logs.json")
AI_REVIEW_EVENT = "ai_review"
MANUAL_RESCUE_EVENT = "manual_rescue"


def _review_key(expression_id: int) -> str:
    return f"expression_review:{expression_id}"


def get_review_state(expression_id: Optional[int]) -> Dict[str, Any]:
    if expression_id is None:
        return {"checked": False, "rejected": False, "modified_by": None}
    value = local_storage[_review_key(expression_id)]
    if isinstance(value, dict):
        return {
            "checked": bool(value.get("checked", False)),
            "rejected": bool(value.get("rejected", False)),
            "modified_by": value.get("modified_by"),
        }
    return {"checked": False, "rejected": False, "modified_by": None}


def set_review_state(
    expression_id: Optional[int],
    checked: bool,
    rejected: bool,
    modified_by: Optional[str],
) -> None:
    if expression_id is None:
        return
    local_storage[_review_key(expression_id)] = {
        "checked": checked,
        "rejected": rejected,
        "modified_by": modified_by,
    }


def _now_iso() -> str:
    return datetime.now().astimezone().isoformat(timespec="seconds")


def _append_log_entry(entry: Dict[str, Any]) -> None:
    entries = read_review_log_entries()
    entries.append(entry)
    REVIEW_LOG_PATH.parent.mkdir(parents=True, exist_ok=True)
    with REVIEW_LOG_PATH.open("w", encoding="utf-8") as log_file:
        json.dump(entries, log_file, ensure_ascii=False, indent=2)


def append_ai_review_log(
    *,
    session_id: str,
    situation: str,
    style: str,
    passed: bool,
    reason: str,
    source: str,
    expression_id: Optional[int] = None,
    error: Optional[str] = None,
) -> Dict[str, Any]:
    """记录一次表达方式 AI 审核结果，供 WebUI 回看和人工救回。"""

    entry: Dict[str, Any] = {
        "id": uuid.uuid4().hex,
        "event": AI_REVIEW_EVENT,
        "created_at": _now_iso(),
        "expression_id": expression_id,
        "session_id": str(session_id or "").strip(),
        "passed": bool(passed),
        "reason": str(reason or "").strip(),
        "situation": str(situation or "").strip(),
        "style": str(style or "").strip(),
        "source": source,
    }
    if error:
        entry["error"] = str(error)

    _append_log_entry(entry)
    logger.debug(
        "表达方式优化记录已写入 "
        f"{REVIEW_LOG_PATH.as_posix()}: passed={entry['passed']} "
        f"session_id={entry['session_id']} reason={entry['reason']}"
    )
    return entry


def append_manual_rescue_log(
    *,
    review_log_id: str,
    expression_id: int,
) -> Dict[str, Any]:
    """记录一次从 AI 审核日志中人工恢复表达方式的操作。"""

    entry: Dict[str, Any] = {
        "id": uuid.uuid4().hex,
        "event": MANUAL_RESCUE_EVENT,
        "created_at": _now_iso(),
        "review_log_id": str(review_log_id),
        "expression_id": int(expression_id),
    }
    _append_log_entry(entry)
    return entry


def read_review_log_entries() -> List[Dict[str, Any]]:
    """读取全部表达方式审核日志，跳过损坏行。"""

    if not REVIEW_LOG_PATH.exists():
        return []

    try:
        with REVIEW_LOG_PATH.open("r", encoding="utf-8") as log_file:
            parsed = json.load(log_file)
    except json.JSONDecodeError as exc:
        logger.warning(f"表达方式审核日志 JSON 解析失败，已忽略当前文件: {exc}")
        return []

    if not isinstance(parsed, list):
        logger.warning("表达方式审核日志格式异常，应为 JSON 数组")
        return []
    return [entry for entry in parsed if isinstance(entry, dict)]


def _parse_created_at(value: Any) -> datetime:
    if isinstance(value, str):
        try:
            return datetime.fromisoformat(value)
        except ValueError:
            return datetime.fromtimestamp(0)
    return datetime.fromtimestamp(0)


def _rescue_by_review_id(entries: List[Dict[str, Any]]) -> Dict[str, Dict[str, Any]]:
    rescue_entries = [
        entry
        for entry in entries
        if entry.get("event") == MANUAL_RESCUE_EVENT and str(entry.get("review_log_id") or "").strip()
    ]
    rescue_entries.sort(key=lambda entry: _parse_created_at(entry.get("created_at")), reverse=True)
    rescues: Dict[str, Dict[str, Any]] = {}
    for rescue_entry in rescue_entries:
        review_log_id = str(rescue_entry.get("review_log_id") or "").strip()
        rescues.setdefault(review_log_id, rescue_entry)
    return rescues


def _with_rescue_state(entry: Dict[str, Any], rescue_entry: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    enriched_entry = dict(entry)
    enriched_entry["rescued"] = rescue_entry is not None
    enriched_entry["rescued_expression_id"] = rescue_entry.get("expression_id") if rescue_entry else None
    enriched_entry["rescued_at"] = rescue_entry.get("created_at") if rescue_entry else None
    return enriched_entry


def get_recent_ai_review_logs(
    *,
    limit: int = 50,
    passed: Optional[bool] = None,
    session_id: Optional[str] = None,
) -> List[Dict[str, Any]]:
    """返回最近的表达方式 AI 审核记录。"""

    entries = read_review_log_entries()
    rescues = _rescue_by_review_id(entries)
    normalized_session_id = str(session_id or "").strip()
    review_entries = [
        entry
        for entry in entries
        if entry.get("event", AI_REVIEW_EVENT) == AI_REVIEW_EVENT and str(entry.get("id") or "").strip()
    ]

    if passed is not None:
        review_entries = [entry for entry in review_entries if bool(entry.get("passed")) is passed]

    if normalized_session_id:
        review_entries = [
            entry
            for entry in review_entries
            if str(entry.get("session_id") or "").strip() == normalized_session_id
        ]

    review_entries.sort(key=lambda entry: _parse_created_at(entry.get("created_at")), reverse=True)
    normalized_limit = max(1, int(limit))
    return [
        _with_rescue_state(entry, rescues.get(str(entry.get("id"))))
        for entry in review_entries[:normalized_limit]
    ]


def get_ai_review_log(review_log_id: str) -> Optional[Dict[str, Any]]:
    """按日志 ID 获取一条表达方式 AI 审核记录。"""

    normalized_id = str(review_log_id or "").strip()
    if not normalized_id:
        return None

    entries = read_review_log_entries()
    rescues = _rescue_by_review_id(entries)
    for entry in entries:
        if entry.get("event", AI_REVIEW_EVENT) != AI_REVIEW_EVENT:
            continue
        if str(entry.get("id") or "").strip() == normalized_id:
            return _with_rescue_state(entry, rescues.get(normalized_id))
    return None
